### PR TITLE
Clean working directory skip paths

### LINKED ISSUE #11099: Let skip paths in cleaning working directory feature
### Feature description

It would be very convenient to clean working directory excluding some exceptional files.

Its label might be **Skip the following path(s)** or **Do not remove following paths**
image_2023-07-11_10-29-09

### Environment

- Git Extensions 4.1.0.16698
- Build 1fe52a137acef1fbd2b653dea263b44ad5844177
- Git 2.40.1.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET 6.0.18
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.18 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

### PR DESCRIPTION
**Pull request by scottsumrall**
Fixes #11099 

## Proposed changes

- Allow users to exclude certain file paths when cleaning repository

### Before

image

### After

image

## Test methodology 

- Manually tested in local repository
- Added and ran relevant unit tests for new functionality

## Test environment(s) 

- GIT 2.41.0 windows.2
- Windows 10

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitExtensions.sln
Comment (RussKie) [inline GitExtensions.sln:357]: Why has this changed?

File: GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs:31]: @merriemcgaw FYI, the new auto-formatting in the designer files is rather unwelcome as it leads to the destructive diffs. It looks like this behaviour is on by default, instead it should be opt-in.

File: GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs:267]: Can you please verify and confirm the tab order is logical across the whole dialog?

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:114]: Could you please clarify why we're prepending `/`? This means root of the drive, not the root of the repo.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:118]: I don't think `.Replace("\\", "/")` should be used like this. Whilst it's not officially supported, the app can run on Linux, so all paths must be in the POSIX form. We have helpers for this specific purpose:
https://github.com/gitextensions/gitextensions/blob/8b5eab531dcc849d00110df6f9797a096e68b500/GitCommands/PathUtil.cs#L19-L24

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:175]: Please invert if and exit early to reduce nesting.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:205]: Ditto

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:195]: These curlies aren't needed.
```cs
using OpenFileDialog openFileDialog = new();
```

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:163]:
```suggestion
            using FolderBrowserDialog dialog = new()
```

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:204]: Doesn't https://learn.microsoft.com/dotnet/api/system.windows.forms.openfiledialog.checkfileexists enforce this?

File: GitUI/CommandsDialogs/FormCleanupRepository.resx
Comment (RussKie) [inline GitUI/CommandsDialogs/FormCleanupRepository.resx:21]: Why has this changed?

File: GitUI/GitExtensionsForm.resx
Comment (RussKie) [inline GitUI/GitExtensionsForm.resx:1]: This file can be removed

File: UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
Comment (RussKie) [inline UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs:510]: Can you please add test cases for multiple includes and excludes too?

**Review by RussKie — CHANGES_REQUESTED** [review]

File: GitExtensions.sln
Comment (scottsumrall) [reply GitExtensions.sln:357]: Mistake on my part, will revert.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: Had some trouble getting the patterns to work, the `/` is definitely not necessary on further testing.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: Good to know, I've implemented it.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: Yep! Fixed it.

File: UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
Comment (scottsumrall) [reply UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs:510]: I'm not sure of the best way to do this. The parameters passed in here for testing, ("paths") and ("excludes"), are functionally equivalent to any amount of paths and excludes as represented in the form of a single string. I could modify the unit tests to use a string parameter like:

(`"\\"path1\\" \\"path2\\""`) 

 to make the formatting explicit if that would be beneficial. I could also just do (`"path1 path2"`).

For now I've gone ahead and removed the "-e /" formatting from the excludes test parameters to bring it in line with the way the include paths test parameter is set up.

File: GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs:267]: Updated the tab order, I'm not positive on best practice in this regard but I think it's good now.

File: GitUI/CommandsDialogs/FormCleanupRepository.resx
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.resx:21]: I'm not sure, this was done automatically. Should I revert the file?

Comment (RussKie) [reply GitUI/CommandsDialogs/FormCleanupRepository.resx:21]: This doesn't look like it's automatic, please revert. Though, if you can positively reproduce it - please file a bug via the VS Feedback button in VS (top right corner).

File: UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
Comment (RussKie) [reply UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs:510]: We can always add new tests for specific scenarios :)

File: GitUI/CommandsDialogs/FormCleanupRepository.resx
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.resx]: I'll look into it. I definitely didn't mess with the .resx files manually, at least that I am aware of. Visual Studio doesn't make it straightforward to modify at that level.

File: GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs:31]: Anything I can/should do to address this?

Comment (gerhardol) [reply GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs:31]: You can manually revert the changes, only applying the actual lines changed.
Simplifies the merge with the other clean PR

**Review by gerhardol — COMMENTED** [review]
looks OK, no deep review, not run

**Review by RussKie — APPROVED** [review]

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:118]: Please use the more meaningful `--exclude` (short option format should be used only by humans when typing the command from command line)

I also would have preferred a better letter than 'a'. 'p' for 'path'?

And maybe use string interpolation instead of string.Format()?

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (pmiossec) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:103]: I would have preferred a better letter than 'a'. 'p' for 'path'?

And maybe use string interpolation instead of string.Format()?

Should we convert to posix path here also?

**Review by pmiossec — CHANGES_REQUESTED** [review]
Please use the more meaningful --exclude option string

**Review by pmiossec — APPROVED** [review]

**Comment by mstv** [comment]
Could you rebase and resolve the conflicts, please?

> Conflicting files
> * GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
> * GitUI/CommandsDialogs/FormCleanupRepository.cs
> * contributors.txt

**Comment by scottsumrall** [comment]
> Could you rebase and resolve the conflicts, please?
> 
> > Conflicting files
> > 
> > * GitUI/CommandsDialogs/FormCleanupRepository.Designer.cs
> > * GitUI/CommandsDialogs/FormCleanupRepository.cs
> > * contributors.txt

Done. Let me know if there are any issues with the rebase, I'm learning as I go here.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:134]:
```suggestion
            // 2. Prepend lines with '--exclude '
```

Same for tests -e there too.

I also would prefer to move this to GitCommandHelpers.cs, just providing the string  or an array (if checkbox set).
(This would affect tests too.)
But I leave this to another maintainer to request, I can live with this.

**Comment by RussKie** [comment]
Judging by the number of commits it doesn't look like you managed to cleanly rebase your work. E.g., there are unrelated commits:
image

Please perform a clean rebase of your work on top of the latest master. https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes#rebase may be of some help.

**Review by RussKie — CHANGES_REQUESTED** [review]
Please rebase

**Comment by scottsumrall** [comment]
> Judging by the number of commits it doesn't look like you managed to cleanly rebase your work. E.g., there are unrelated commits: image
> 
> Please perform a clean rebase of your work on top of the latest master. https://github.com/gitextensions/gitextensions/wiki/How-To%3A-Squash-and-Rebase-your-changes#rebase may be of some help.

Apologies for the mix up, still figuring out the git workflow. Please let me know if this is looking correct. 
image
Additionally, are you saying I should squash these commits together?

**Comment by gerhardol** [comment]
It is not fixed. Rebase preserving commits is not straight forward as there are intermediate conflicts.
 
In this case, it is easiest to squash on the BASE (soft reset), commit with proper message and then rebase on master. There are quite some commits here, I do not think they add value in a further review (review has basically to be restarted, in this case it is OK).
There should only be 6 changed files in the commit, other files have not been changed in master.

**Comment by scottsumrall** [comment]
> It is not fixed. Rebase preserving commits is not straight forward as there are intermediate conflicts.
> 
> In this case, it is easiest to squash on the BASE (soft reset), commit with proper message and then rebase on master. There are quite some commits here, I do not think they add value in a further review (review has basically to be restarted, in this case it is OK). There should only be 6 changed files in the commit, other files have not been changed in master.

Would you mind taking a look at this temp branch I made? I'd like to make sure I'm doing this right before I start overwriting things here.
https://github.com/gitextensions/gitextensions/compare/master...scottsumrall:gitextensions:skip-paths-test

To get this PR's branch to the same state I should just need to force push then do a simple rebase for the latest changes.

**Comment by gerhardol** [comment]
> Would you mind taking a look at this temp branch I made? I'd like to make sure I'm doing this right before I start overwriting things here. [master...scottsumrall:gitextensions:skip-paths-test](https://github.com/gitextensions/gitextensions/compare/master...scottsumrall:gitextensions:skip-paths-test)

Fine, except that GitCommandHelpersTest.cs still has -e instead of --exclude
Hope you learn something!

**Comment by scottsumrall** [comment]
Appreciate the assistance!

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
@scottsumrall thank you.

@gerhardol please merge, if all your feedback is addressed.

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (mstv) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:65]: ❔ Should we apply `excludePathArguments` for submodules, too?

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (mstv) [inline GitUI/CommandsDialogs/FormCleanupRepository.cs:137]: Nit: remove the empty line, please

**Review by mstv — COMMENTED** [review]
LGTM, have not run

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: changed in one of my commits

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: I see two viewpoints:
* Paths are relative the superproject and only should be relative the master. Hard to implement though, if wildcards are to be allowed.
* apply the same as for the superproject, similar to include paths. Easier to implement, maybe not expected?

Probably second option.
@scottsumrall thoughts?

**Comment by gerhardol** [comment]
I pushed some commits, the merge was bad and preview removed files.
Part of that was that use of common code I suggsested in #11101 was not done, so I added it again.
Also using string interpolation instead of string.format - pushed a change instead of suggesting (also for existing code)

**Comment by gerhardol** [comment]
@scottsumrall Please acknowledge the changes and I will merge

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (scottsumrall) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: I'm not sure I completely follow. As of right now, the excludes for the general clean are only for excluding individual files and do not support wildcards. I assume it would be the same for submodules? If we want this feature included I would lean towards the first solution but the second option seems fine for its current state.

**Comment by scottsumrall** [comment]
Looks good to me

File: GitUI/CommandsDialogs/FormCleanupRepository.cs
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormCleanupRepository.cs]: The first option would affect the include files too. But that wasrecently added without consideration.
However, I do not find it reasonable.
Use second alternative, in this PR before merge.

Add exclude paths to `CleanSubmodules()`
(Make sure there is a test as well, also for include not tested at all now)

**Comment by scottsumrall** [comment]
@gerhardol Should I just open a new PR with the changes we discussed or should I create a new issue?

**Comment by gerhardol** [comment]
> @gerhardol Should I just open a new PR with the changes we discussed or should I create a new issue?

A new PR only is enough. PR not needed regardless - not big enough to track in my opinion, still reasonable to make this change 'complete'.

**Pull request merged**
